### PR TITLE
Advanced Travis integration

### PR DESCRIPTION
**Pull request by edwinmp**
Add selenium web driver tests to Travis

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,19 +1,39 @@
 sudo: required
 dist: trusty
-
 language: node_js
 node_js:
-  - "6"
-
+- '6'
```
Comment (acellam) [inline .travis.yml]: any reasons for single quotes ? use double quotes

File: .travis.yml [hunk 2]
```diff
@@ -1,19 +1,41 @@
...
-  - export DISPLAY=:99.0
-  - sh -e /etc/init.d/xvfb start
-
+- npm i -g npm
+- npm install grunt-cli -g
+- export CHROME_BIN=chromium-browser
+- export DISPLAY=:99.0
+- sh -e /etc/init.d/xvfb start
+- wget -q -O - http://opensource.wandisco.com/wandisco-debian.gpg | sudo apt-key add
+  -
```
Comment (acellam) [inline .travis.yml]: Remove `-`

**Review by acellam — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 2 above)
Comment (edwinmp) [reply .travis.yml]: Damn, this keeps happening automatically for some unknown reason. Let me try again.

File: .travis.yml (hunk 1 above)
Comment (edwinmp) [reply .travis.yml]: Changed.

File: src/RichText.webmodeler.ts [hunk 3]
```diff
@@ -6,50 +6,23 @@ import { RichTextContainerProps } from "./components/RichTextContainer";
...
+        return createElement(ValidateConfigs, { ...this.props as RichTextContainerProps, showOnError: true },
+            createElement(RichText, preview.transformProps(this.props))
+        );
     }
 
     private static transformProps(props: RichTextContainerProps): RichTextProps {
         const valueAttribute = props.stringAttribute ? props.stringAttribute.split(".")[ 2 ] : "";
 
         return {
+            ... props as any,
```
Comment (Andries-Smit) [inline src/RichText.webmodeler.ts]: Think any type should be solved differently.
Note that `hasContext` is not a prop of RichText

Comment (edwinmp) [reply src/RichText.webmodeler.ts]: Removed. Explicitly adding the props.

File: .travis.yml [hunk 4]
```diff
@@ -1,19 +1,40 @@
...
 cache:
   directories:
-    - "node_modules"
+  - node_modules
+env:
+  global:
+  - MX_APP_NAME=texteditorwidget
+  - MX_BRANCH_NAME=trunk
+  - MX_PROJECT_ID=8648ccda-9281-4768-abb2-8ce61a80e2f1
+  - secure: CvE9K9XGm1GD0tTM38cHX1/OYhgvgHZTHbIs+PMGiSb4FVATN0BRLUay3IqvoohnxKcBtUbSGqCZBZ6CNRxqdcq0KRX83n+bPibD2XOipEvW6SOOiiF91qV1qTJqfbPXn+lsshZihrYAS1aBfEeqhWniVcqPlnsC20ypYBx8TI2jMMylKQfb+D8tXX2XkY+rlZcFIA0tDJ602hF29bN79mzTxh/zDO8+0DHpvghJtTFDJrWCrPUrX5vSXy+FQR3QFkhRTD7n9Y5F8OTe4WZMkhWyYj9eV6Agf1JkUZK4XT82N+3742Zzf9O/O5IeBQlWC3Vn+NqujKc6pvcui8iEH0Mm2PRGvXUxlvNLwI4sPDILO7av8TTEQY4ejrjBzwf+UJFjpBgzgTm/L8+FXsqlvqr7C3G3aHLXyCsmLF4rOU9/m5woLbOMZ4l3h0cl5YIYKCEEEQzKoyQza/yUEXm0vMGXqNI29iQXI+bu2YBB006rtOGvi8mdVJZmotuvNXmbFV6SsMPCaKX+/OIZ70zU17xdQSBrCFPLyVEhCxjR6So5KBG/EXPRq3RHtfsJsxNsxK4n1HkhocrY0snmiXwRW8YMLtBHdtU6z+uOngfogAKRD7FkdCwMYBsomP4EeyA5uXDWyT9bQcjsyi/OAJbHoEZdeL5X/aOgxUeOqPBWY68=
```
Comment (Andries-Smit) [inline .travis.yml]: Secure keys should no longer be in the yml, they should be environment setting

File: src/components/RichText.ts [hunk 5]
```diff
@@ -143,6 +143,22 @@ export class RichText extends Component<RichTextProps, {}> {
         }
     }
 
+    private handleTextChange() {
+        if (this.quill && this.quill.hasFocus() && this.props.onChange) {
+            const value = this.quill.root.innerHTML;
+            if (this.props.value !== value) {
+                this.props.onChange(value);
+                this.textChanged = true;
+            }
```
Comment (Andries-Smit) [inline src/components/RichText.ts]: missing else construction. `esle { this.textChanged = false; }`

File: src/components/RichTextContainer.ts [hunk 6]
```diff
@@ -83,21 +105,38 @@ export default class RichTextContainer extends Component<RichTextContainerProps,
...
+
+            this.props.mxform.listen("submit", this.onFormSubmit);
         }
     }
 
     private handleSubscriptions() {
         this.setState({
             value: getValue(this.props.stringAttribute, "", this.props.mxObject) as string
-        });
+        }, () => this.update = true);
```
Comment (Andries-Smit) [inline src/components/RichTextContainer.ts]: Why is `this.update` not part of `this.state`?

File: package.json [hunk 7]
```diff
@@ -23,18 +32,19 @@
...
-    "@types/enzyme": "^2.8.2",
+    "@types/classnames": "^2.2.1",
+    "@types/core-js": "^0.9.42",
+    "@types/enzyme": "^2.8.6",
     "@types/jasmine": "^2.5.53",
     "@types/karma": "^0.13.36",
     "@types/quill": "0.0.31",
-    "@types/react": "^15.0.38",
-    "@types/react-dom": "^15.5.1",
+    "@types/react": "^16.0.2",
```
Comment (Andries-Smit) [inline package.json]: type should be 16. as the web modeler is not supporting it. risking to use react 16 features

File: src/RichText.webmodeler.ts [hunk 8]
```diff
@@ -1,55 +1,26 @@
...
     private static transformProps(props: RichTextContainerProps): RichTextProps {
         const valueAttribute = props.stringAttribute ? props.stringAttribute.split(".")[ 2 ] : "";
 
         return {
-            className: props.class,
-            customOptions: props.customOptions,
-            editorOption: props.editorOption,
-            maxNumberOfLines: props.maxNumberOfLines,
-            minNumberOfLines: props.minNumberOfLines,
+            ... RichTextContainer.getPartialRichTextProps(props) as RichTextProps,
```
Comment (Andries-Smit) [inline src/RichText.webmodeler.ts]: Personally, no so fan of this solution. as you need to move the an other file to see which props are copied.

File: src/components/RichText.ts [hunk 9]
```diff
@@ -20,27 +20,35 @@ export interface CommonRichTextProps {
...
+    recreate?: boolean;
 }
 
 export interface RichTextProps extends CommonRichTextProps {
     className?: string;
-    onChange?: (value: string) => void;
+    onChange?: (value?: string) => void;
+    onBlur?: () => void;
+    update: boolean;
+    getQuill?: (quill: Quill.Quill) => void;
```
Comment (Andries-Smit) [inline src/components/RichText.ts]: Function is still reference but is not used actively in the container

File: src/components/RichText.ts [hunk 10]
```diff
@@ -20,27 +20,35 @@ export interface CommonRichTextProps {
...
     maxNumberOfLines: number;
+    recreate?: boolean;
 }
 
 export interface RichTextProps extends CommonRichTextProps {
     className?: string;
-    onChange?: (value: string) => void;
+    onChange?: (value?: string) => void;
+    onBlur?: () => void;
+    update: boolean;
```
Comment (Andries-Smit) [inline src/components/RichText.ts]: I don't understand `update` is there a better name, and why is it needed, it makes the code not react like and feels hackish (what is probably is, and it that case i expect some comment in the code)

**Review by Andries-Smit — CHANGES_REQUESTED** [review]

File: src/components/RichText.ts (hunk 9 above)
Comment (edwinmp) [reply src/components/RichText.ts]: Removed.

File: src/RichText.webmodeler.ts (hunk 8 above)
Comment (edwinmp) [reply src/RichText.webmodeler.ts]: Inlined them all. Same in the container.

File: package.json (hunk 7 above)
Comment (edwinmp) [reply package.json]: Downgraded to version 15.0.28

File: src/components/RichText.ts (hunk 10 above)
Comment (edwinmp) [reply src/components/RichText.ts]: Renamed to updateEditor and comment added as well.

File: src/components/RichTextContainer.ts (hunk 6 above)
Comment (edwinmp) [reply src/components/RichTextContainer.ts]: It's purpose wouldn't be served in the state, so this was the only option. Purpose was to control update behaviour between changes triggered by the on change feature & those made externally (changing the value of the attribute outside the widget).

File: .travis.yml [hunk 11]
```diff
@@ -1,19 +1,40 @@
...
+deploy:
+  provider: releases
+  api_key:
+    secure: jkZxWr+Crfkvth5sYDuX3r/TAjGW4YTddRe41jHu9Znubo2wM7i0qAK3ByoWEQMJdyFZbYhNQVXV5JpsXH+7/NlNkOHb7SLPqrWhVGsBNSUjzWXXrJp6TJQlc1PMh/DwGxWG/Q6TQ1SyHVTAEDoi+VDXD4jV1d5pGNXrA7D/ShN5eLGs7sktl9KhBAFKfABkBGHnz1QLw8AFvtLGI7iSOVMpOhbSAxLj20NYR+cBhgx7DPGJ5Cnp5Huhz7ky75WPtdd79H0bWCZFyE01BuDnsm5Bp/W87Pg9vsIaD1WWdfdrzucxFVSaKsjC42I5/ILretlPl0Y+UO/Km3MZIch6YfIB+p7TrIn4R6qXTKdnFtW40R0ziu2UpcFCW2KV0P0rOZojPs9u/r2pxJdz/b0UDQP5hMwJoj/0R2V9VLHs24kUSystmu1wKSJVC9j74RhZJxVzO/SAUtfULHqLqINrd9ewcct163KzTq4blVDHn/5F/2Lj5r/Gkn3Qp9tdjPBwntVt46Z+tFK2fuwZ6RNkyHJVsQcyu0OS8YYxLHW1vAmK7qiwANcywn2tg7Q8KflZE04yqh3Buv1x9cLoS5BiUncS+YXuPH0M/u8pUR/fc6EEofo0PhYlndn2/xrnniunD+VYcjtD2aMjCSiHz2wmGLQGT5VOqL0xblDcxQASVvg=
+  file_glob: true
+  file: dist/release/*
+  skip_cleanup: true
+  on:
+    repo: FlockOfBirds/rich-text
```
Comment (Andries-Smit) [inline .travis.yml]: dont hard code your repo, its available in environment variable ` repo: $TRAVIS_REPO_SLUG`

Comment (edwinmp) [reply .travis.yml]: Done.

File: .travis.yml (hunk 4 above)
Comment (edwinmp) [reply .travis.yml]: Done.
